### ISSUE TITLE
Websocket Client disconnects when recieving large amounts of data

### ISSUE THREAD

**Issue by eWert-Online** [anmonteiro/httpun-ws#23]
I am using this library to create a websocket client, which connects to a running browser and controls it through the Chrome Devtools Protocol. 

While [capturing a screenshot](https://chromedevtools.github.io/devtools-protocol/tot/Page/#method-captureScreenshot) of a webpage, the Browser tries to send the base64 encoded image through the websocket connection, which leads to the client disconnecting.

It works with data which has a frame length of less than 4096. I don't know if this is just a coincidence though.

**Comment by anmonteiro** [comment]
The 4096 byte pointer sounds helpful. Have you tried configuring a bigger buffer size?

**Comment by eWert-Online** [comment]
No, but I am not really sure, where I would be able to do that, as I don't have any useage of the Buffer module in my code. 
(I am also quite new to native reason / ocaml)

The respone actually doesnt even make into my websocket_handler or any part of my code before it gets closed.

**Comment by anmonteiro** [comment]
This library does unbuffered parsing from the read buffer, which is limited to 4096 bytes in size right now.

The immediate workaround is to allow you to configure that, which might not be available now.

Otherwise we should probably explore fixing parsing not to have this limitation, but that might require some more thinking.

Could you paste an example of how you're setting up this library in your code? Which APIs are you using?

**Comment by eWert-Online** [comment]
Sure!
Here is the code I am using to setup a connection to the Chrome Devtools Protocol. 
It is largely based on the wscat example.

<details>
  <summary>Code</summary>
    
```re
let close_requests = Queue.create();
let pending_requests = Queue.create();
let sent_requests = Hashtbl.create(10);

let websocket_handler = (u, wsd) => {
  let rec input_loop = wsd => {
    let%lwt () = Lwt.pause();
    if (!Queue.is_empty(close_requests)) {
      Websocketaf.Wsd.close(wsd);
      close_requests |> Queue.iter(resolver => Lwt.wakeup_later(resolver, ()));
      close_requests |> Queue.clear;
      Lwt.return_unit;
    } else if (!Queue.is_empty(pending_requests)) {
      let (key, message, resolver) = Queue.take(pending_requests);
      print_endline("[SOCKET] sending: " ++ message);
      let payload = Bytes.of_string(message);
      Websocketaf.Wsd.send_bytes(
        wsd,
        ~kind=`Text,
        payload,
        ~off=0,
        ~len=Bytes.length(payload),
      );
      Hashtbl.add(sent_requests, key, resolver);
      input_loop(wsd);
    } else {
      input_loop(wsd);
    };
  };
  Lwt.async(() => input_loop(wsd));

  let frame = (~opcode, ~is_fin, bs, ~off, ~len) => {
    print_newline();
    print_newline();
    print_int(opcode |> Websocketaf.Websocket.Opcode.code);
    print_newline();
    print_endline(
      "[SOCKET] GETTING DATA FIN:" ++ (is_fin ? "true" : "false"),
    );
    print_endline("[SOCKET] GETTING DATA LEN:" ++ string_of_int(len));
    let payload = Bytes.create(len);
    Lwt_bytes.blit_to_bytes(bs, off, payload, 0, len);
    let response = Bytes.to_string(payload);
    print_endline("[SOCKET] got: " ++ response);
    let id =
      response
      |> Yojson.Safe.from_string
      |> Yojson.Safe.Util.member("id")
      |> Yojson.Safe.Util.to_int_option;
    switch (id) {
    | None => ()
    | Some(key) =>
      let resolver = Hashtbl.find(sent_requests, key);
      Lwt.wakeup_later(resolver, response);
      Hashtbl.remove(sent_requests, key);
    };
  };

  let eof = () => {
    print_endline("[CLOSING SOCKET]");
    Lwt.wakeup_later(u, ());
  };

  Websocketaf.Client_connection.{frame, eof};
};

let error_handler = error => {
  print_endline("[SOCKET] GOT ERROR");
  switch (error) {
  | `Exn(_exn) => ()
  | `Handshake_failure(_resp, _body) => ()
  | `Invalid_response_body_length(_resp) => ()
  | `Malformed_response(_string) => ()
  | _ => assert(false)
  };
};

let connect = url => {
  let uri = Uri.of_string(url);
  let port = uri |> Uri.port |> Option.value(~default=0);
  let host = uri |> Uri.host_with_default(~default="");
  let resource = uri |> Uri.path;

  let%lwt addresses =
    Lwt_unix.getaddrinfo(
      host,
      string_of_int(port),
      [Unix.(AI_FAMILY(PF_INET))],
    );
  let socket = Lwt_unix.socket(Unix.PF_INET, Unix.SOCK_STREAM, 0);
  let%lwt () = Lwt_unix.connect(socket, List.hd(addresses).Unix.ai_addr);

  let (p, u) = Lwt.wait();

  let _ =
    socket
    |> Websocketaf_lwt_unix.Client.connect(
         ~nonce="0123456789ABCDEF",
         ~host,
         ~port,
         ~resource,
         ~error_handler,
         ~websocket_handler=websocket_handler(u),
       );
  p;
};

let send = message => {
  let key =
    message
    |> Yojson.Safe.from_string
    |> Yojson.Safe.Util.member("id")
    |> Yojson.Safe.Util.to_int;

  let (p, resolver) = Lwt.wait();
  pending_requests |> Queue.add((key, message, resolver));
  p;
};

let close = () => {
  let (p, resolver) = Lwt.wait();
  close_requests |> Queue.add(resolver);
  p;
};
```
  
</details>

What do you mean by _"Which APIs are you using?"_?  To which ws server I am trying to connect?
Then the answer would be a running headless chromium instance through the [Chrome DevTools Protocol](https://chromedevtools.github.io/devtools-protocol/)
You may also look at the repo here: https://github.com/eWert-Online/osnap

**Comment by anmonteiro** [comment]
> Sure!
> Here is the code I am using to setup a connection to the Chrome Devtools Protocol.
> It is largely based on the wscat example.
> 
> Code

great, this is what I was asking for. Thanks!


I'll take a look at this soon, stay tuned.

**Comment by anmonteiro** [comment]
Hey @eWert-Online, I fixed this in #27. You can now pass an additional `~config` argument to `Websocketaf_lwt_unix.Client.create`. We'll use `read_buffer_size` to configure the max buffer size for payloads.

Additionally, #25 is a work in progress that will allow us to have a buffered reader (allowing us to read larger payloads), without the need for configuring a big read buffer size.
